docs/TypeScript.md

```diff
@@ -1,291 +1,463 @@
 
-This code shows the typings for the `fastify-static` plugin.
+#### fastify.RouteHandlerMethod
 
-```ts
-// require fastify typings
-import fastify = require("fastify");
-// require necessary http typings
-import { Server, IncomingMessage, ServerResponse } from "http";
+#### fastify.RouteOptions
 
-// extend fastify typings
-declare module "fastify" {
-    interface FastifyReply<HttpResponse> {
-        sendFile(filename: string): FastifyReply<HttpResponse>;
-    }
-}
+#### fastify.RouteShorthandMethod
 
-// declare plugin type using fastify.Plugin
-declare const fastifyStatic: fastify.Plugin<Server, IncomingMessage, ServerResponse, {
-    root: string;
-    prefix?: string;
-    serve?: boolean;
-    decorateReply?: boolean;
-    schemaHide?: boolean;
-    setHeaders?: (...args: any[]) => void;
-    acceptRanges?: boolean;
-    cacheControl?: boolean;
-    dotfiles?: boolean;
-    etag?: boolean;
-    extensions?: string[];
-    immutable?: boolean;
-    index?: string[];
-    lastModified?: boolean;
-    maxAge?: string | number;
-}>;
-
-// export plugin type
-export = fastifyStatic;
-```
+#### fastify.RouteShorthandOptions
 
-Now you are good to go and could use the plugin like so:
-
-```ts
-import * as Fastify from 'fastify'
-import * as fastifyStatic from 'fastify-static'
-
-const app = Fastify()
-
-// the options here are type-checked
-app.register(fastifyStatic, {
-  acceptRanges: true,
-  cacheControl: true,
-  decorateReply: true,
-  dotfiles: true,
-  etag: true,
-  extensions: ['.js'],
-  immutable: true,
-  index: ['1'],
-  lastModified: true,
-  maxAge: '',
-  prefix: '',
-  root: '',
-  schemaHide: true,
-  serve: true,
-  setHeaders: (res, pathName) => {
-    res.setHeader('some-header', pathName)
-  }
-})
+#### fastify.RouteShorthandOptionsWithHandler
 
-app.get('/file', (request, reply) => {
-  // using newly defined function on FastifyReply
-  reply.sendFile('some-file-name')
-})
-```
+#### fastify.RegisterOptions
+
+#### fastify.FastifyBodyParser
+
+#### fastify.FastifyContentTypeParser
+
+#### fastify.AddContentTypeParser
+
+#### fastify.hasContentTypeParser
+
+#### fastify.FastifyError
+
+#### fastify.ValidationResult
+
+#### fastify.FastifySchema
+
+#### fastify.FastifySchemaCompiler
+
+#### fastify.HTTPMethods
+
+#### fastify.RawServerBase
+
+#### fastify.RawRequestDefaultExpression
+
+#### fastify.RawReplyDefaultExpression
+
+#### fastify.RawServerDefault
+
+#### fastify.onCloseHookHandler
+
+#### fastify.onRouteHookHandler
+
+#### fastify.onRequestHookHandler
+
+#### fastify.onSendHookHandler
+
+#### fastify.onErrorHookHandler
+
+#### fastify.preHandlerHookHandler
+
+#### fastify.preParsingHookHandler
+
+#### fastify.preSerializationHookHandler
+
+#### fastify.preValidationHookHandler
+
+#### fastify.AddHook
+
+#### fastify.addHookHandler
+
+#### fastify.FastifyServerFactory
 
-Adding typings to all our plugins is a community effort so feel free to contribute!
+#### fastify.FastifyServerFactoryHandler
```

Comment:
Do we need all those empty blocks?

Reply:
Haven't filled them in yet. Probably wont even have to document all of it; theres just a couple specific cases I want to make sure to highlight because the generics can get a little complicated.